swayfmt/src/formatter/shape.rs

```diff
@@ -252,68 +245,67 @@ impl Shape {
                 } else {
-                    self.code_line.update_line_style(LineStyle::Normal)
+                    self.code_line.with_line_style(LineStyle::Normal)
                 }
             }
             ExprKind::Import => {
-                if self.width > config.whitespace.max_width {
-                    self.code_line.update_line_style(LineStyle::Multiline)
+                if self.code_line.width > config.whitespace.max_width {
+                    self.code_line.with_line_style(LineStyle::Multiline)
                 } else {
-                    self.code_line.update_line_style(LineStyle::Normal)
+                    self.code_line.with_line_style(LineStyle::Normal)
                 }
             }
             ExprKind::Function => {
-                if self.width > config.whitespace.max_width
+                if self.code_line.width > config.whitespace.max_width
                     || body_width.unwrap_or(0) > self.width_heuristics.fn_call_width
                 {
-                    self.code_line.update_line_style(LineStyle::Multiline)
+                    self.code_line.with_line_style(LineStyle::Multiline)
                 } else {
-                    self.code_line.update_line_style(LineStyle::Normal)
+                    self.code_line.with_line_style(LineStyle::Normal)
                 }
             }
             ExprKind::Conditional => {
-                if self.width < self.width_heuristics.single_line_if_else_max_width {
-                    self.code_line.update_line_style(LineStyle::Inline)
+                if self.code_line.width < self.width_heuristics.single_line_if_else_max_width {
+                    self.code_line.with_line_style(LineStyle::Inline)
                 } else if body_width.unwrap_or(0) > self.width_heuristics.chain_width {
-                    self.code_line.update_line_style(LineStyle::Multiline)
+                    self.code_line.with_line_style(LineStyle::Multiline)
                 } else {
-                    self.code_line.update_line_style(LineStyle::Normal)
+                    self.code_line.with_line_style(LineStyle::Normal)
                 }
             }
-            _ => self.code_line.update_line_style(LineStyle::default()),
+            _ => self.code_line.with_line_style(LineStyle::default()),
         }
     }
-    /// Update `Shape::code_line` with the provided settings.
-    pub(crate) fn update_line_settings(&mut self, line_settings: CodeLine) {
-        self.code_line = line_settings
-    }
-    /// Reset `Shape::code_line` to default.
-    pub(crate) fn reset_line_settings(&mut self) {
-        self.update_line_settings(CodeLine::default())
+    /// Create a new `Shape` with a new `CodeLine` from a given `LineStyle` and `ExprKind`.
+    pub(crate) fn with_code_line_from(self, line_style: LineStyle, expr_kind: ExprKind) -> Self {
+        Self {
+            code_line: CodeLine::from(line_style, expr_kind),
```

Comment:
Note that the returned `Shape`'s `code_line` here will not have the original `self.code_line`'s `width` or `has_where_clause` (both are set to `default()` in `CodeLine::from`). Is that definitely what we want here? Or should the returned `Shape` have the original `Shape`'s `code_line.width` and `code_line.has_where_clause`?

Reply:
Hm, I added an option for them previously while I was building the first iteration of this new API, and it seemed that they were always `width: 0` and `has_where_clause: None`, and I think this is the right move. Each time we are addressing a new `CodeLine` we want to have a fresh start although sometimes we already know some information like `ExprKind` or in the case of say `if/else` we already know it's a `Conditional` so we give that information at the start. We never know the `width` before analyzing the `CodeLine` nor if it has a `where` clause so I think this is the safest route, at least for now.

Reply:
I'm open ears for a better approach!